db/schema_tables.cc

```diff
@@ -958,7 +958,7 @@ mutation compact_for_schema_digest(const mutation& m) {
     // See https://issues.apache.org/jira/browse/CASSANDRA-6862.
     // We achieve similar effect with compact_for_compaction().
     mutation m_compacted(m);
-    m_compacted.partition().compact_for_compaction(*m.schema(), always_gc, gc_clock::time_point::max());
+    m_compacted.partition().compact_for_compaction(*m.schema(), always_gc, m.decorated_key(), gc_clock::time_point::max());
```

Comment:
I think we can make the drop tombstones unconditionally intent explicit and not play games with always gc and max timestamp.  

I added something like below:

```
commit 5979018e3fc2a9ce6f2256ffd81ed070476789ad (HEAD -> repair_gc_protection)
Author: Asias He <asias@scylladb.com>
Date:   Fri Dec 3 09:08:40 2021 +0800

    mutation_partition: Add compact_for_compaction_drop_tombstones_unconditionally

diff --git a/mutation_partition.hh b/mutation_partition.hh
index 777761e004..d11fecc3f0 100644
--- a/mutation_partition.hh
+++ b/mutation_partition.hh
@@ -1245,7 +1245,8 @@ class mutation_partition final {
         bool always_return_static_content,
         bool reverse,
         uint64_t row_limit,
-        can_gc_fn&);
+        can_gc_fn&,
+        bool drop_tombstones_unconditionally);
 
     // Calls func for each row entry inside row_ranges until func returns stop_iteration::yes.
     // Removes all entries for which func didn't return stop_iteration::no or wasn't called at all.
@@ -1285,6 +1286,10 @@ class mutation_partition final {
         const dht::decorated_key& dk,
         gc_clock::time_point compaction_time);
 
+    // Like compact_for_compaction but drop tombstones unconditionally
+    void compact_for_compaction_drop_tombstones_unconditionally(const schema& s,
+            const dht::decorated_key& dk);
+
     // Returns the minimal mutation_partition that when applied to "other" will
     // create a mutation_partition equal to the sum of other and this one.
     // This and other must both be governed by the same schema s.
diff --git a/db/schema_tables.cc b/db/schema_tables.cc
index 2309642c9c..fd0d4b3962 100644
--- a/db/schema_tables.cc
+++ b/db/schema_tables.cc
@@ -958,7 +958,7 @@ mutation compact_for_schema_digest(const mutation& m) {
     // See https://issues.apache.org/jira/browse/CASSANDRA-6862.
     // We achieve similar effect with compact_for_compaction().
     mutation m_compacted(m);
-    m_compacted.partition().compact_for_compaction(*m.schema(), always_gc, m.decorated_key(), gc_clock::time_point::max());
+    m_compacted.partition().compact_for_compaction_drop_tombstones_unconditionally(*m.schema(), m.decorated_key());
     return m_compacted;
 }
 
diff --git a/mutation_partition.cc b/mutation_partition.cc
index eb44793bab..88b3f72779 100644
--- a/mutation_partition.cc
+++ b/mutation_partition.cc
@@ -1297,12 +1297,14 @@ uint32_t mutation_partition::do_compact(const schema& s,
     bool always_return_static_content,
     bool reverse,
     uint64_t row_limit,
-    can_gc_fn& can_gc)
+    can_gc_fn& can_gc,
+    bool drop_tombstones_unconditionally)
 {
     check_schema(s);
     assert(row_limit > 0);
 
-    auto gc_before = ::get_gc_before_for_key(s.shared_from_this(), dk, query_time);
+    auto gc_before = drop_tombstones_unconditionally ? gc_clock::time_point::max() :
+        ::get_gc_before_for_key(s.shared_from_this(), dk, query_time);
 
     auto should_purge_tombstone = [&] (const tombstone& t) {
         return t.deletion_time < gc_before && can_gc(t);
@@ -1369,7 +1371,8 @@ mutation_partition::compact_for_query(
     uint64_t row_limit)
 {
     check_schema(s);
-    return do_compact(s, dk, query_time, row_ranges, always_return_static_content, reverse, row_limit, always_gc);
+    bool drop_tombstones_unconditionally = false;
+    return do_compact(s, dk, query_time, row_ranges, always_return_static_content, reverse, row_limit, always_gc, drop_tombstones_unconditionally);
 }
 
 void mutation_partition::compact_for_compaction(const schema& s,
@@ -1380,7 +1383,19 @@ void mutation_partition::compact_for_compaction(const schema& s,
         query::clustering_range::make_open_ended_both_sides()
     };
 
-    do_compact(s, dk, compaction_time, all_rows, true, false, query::partition_max_rows, can_gc);
+    bool drop_tombstones_unconditionally = false;
+    do_compact(s, dk, compaction_time, all_rows, true, false, query::partition_max_rows, can_gc, drop_tombstones_unconditionally);
+}
+
+void mutation_partition::compact_for_compaction_drop_tombstones_unconditionally(const schema& s, const dht::decorated_key& dk)
+{
+    check_schema(s);
+    static const std::vector<query::clustering_range> all_rows = {
+        query::clustering_range::make_open_ended_both_sides()
+    };
+    bool drop_tombstones_unconditionally = true;
+    auto compaction_time = gc_clock::time_point::max();
+    do_compact(s, dk, compaction_time, all_rows, true, false, query::partition_max_rows, always_gc, drop_tombstones_unconditionally);
 }
 
```

Reply:
> Generally, I think we should keep mutation_partition -level algorithms global-state-free to avoid surprises.

How would this work if mutation_partition still use gc_grace_seconds to gc?

Reply:
I'm suggesting that it doesn't use gc_grace_seconds directly, but rely on externally provided can_gc. We already pass can_gc, we can just make it responsible for more decisions (to include the gc_grace_seconds decision).

Reply:
> I think we can make the drop tombstones unconditionally intent explicit and not play games with always gc and max timestamp.

> I added something like below:

That would work too, however it makes algorithms more complex.

Reply:
can_gc does not express we want to drop tombstones unconditionally. I think it is more clear to have a compact_for_compaction_drop_tombstones_unconditionally().

Reply:
can_gc does not express this currently, but it can be made to express it.

Reply:
Still, it is not clear than compact_for_compaction_drop_tombstones_unconditionally IMO.